File: backend/domain/workflow/variable/variable.go
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package variable

import (
	"context"
	"fmt"

	"github.com/cloudwego/eino/compose"

	"github.com/coze-dev/coze-studio/backend/domain/workflow/entity/vo"
)

var variableHandlerSingleton *Handler

func GetVariableHandler() *Handler {
	return NewVariableHandler()
}

type Handler struct {
	UserVarStore   Store
	SystemVarStore Store
	AppVarStore    Store
}

func (v *Handler) Get(ctx context.Context, t vo.GlobalVarType, path compose.FieldPath, opts ...OptionFn) (any, error) {
	switch t {
	case vo.GlobalUser:
		return v.UserVarStore.Get(ctx, path, opts...)
	case vo.GlobalSystem:
		return v.SystemVarStore.Get(ctx, path, opts...)
	case vo.GlobalAPP:
		return v.AppVarStore.Get(ctx, path, opts...)
	default:
		return nil, fmt.Errorf("unknown variable type: %v", t)
	}
}

func (v *Handler) Set(ctx context.Context, t vo.GlobalVarType, path compose.FieldPath, value any, opts ...OptionFn) error {
	switch t {
	case vo.GlobalUser:
		return v.UserVarStore.Set(ctx, path, value, opts...)
	case vo.GlobalSystem:
		return v.SystemVarStore.Set(ctx, path, value, opts...)
	case vo.GlobalAPP:
		return v.AppVarStore.Set(ctx, path, value, opts...)
	default:
		return fmt.Errorf("unknown variable type: %v", t)
	}
}

func (v *Handler) Init(ctx context.Context) context.Context {
	if v.UserVarStore != nil {
		v.UserVarStore.Init(ctx)
	}

	if v.SystemVarStore != nil {
		v.SystemVarStore.Init(ctx)
	}

	if v.AppVarStore != nil {
		v.AppVarStore.Init(ctx)
	}

	return ctx
}

type StoreInfo struct {
	AppID        *int64
	AgentID      *int64
	ConnectorID  int64
	ConnectorUID string
}

type StoreConfig struct {
	StoreInfo StoreInfo
}

type OptionFn func(*StoreConfig)

func WithStoreInfo(info StoreInfo) OptionFn {
	return func(option *StoreConfig) {
		option.StoreInfo = info
	}
}

//go:generate mockgen -destination varmock/var_mock.go --package mockvar -source variable.go
type Store interface {
	Init(ctx context.Context)
	Get(ctx context.Context, path compose.FieldPath, opts ...OptionFn) (any, error)
	Set(ctx context.Context, path compose.FieldPath, value any, opts ...OptionFn) error
}

var variablesMetaGetterImpl VariablesMetaGetter

func GetVariablesMetaGetter() VariablesMetaGetter {
	return NewVariablesMetaGetter()
}

type VariablesMetaGetter interface {
	GetAppVariablesMeta(ctx context.Context, id, version string) (m map[string]*vo.TypeInfo, err error)
	GetAgentVariablesMeta(ctx context.Context, id int64, version string) (m map[string]*vo.TypeInfo, err error)
}

File: frontend/packages/workflow/feature-encapsulate/src/utils/variable.ts
/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { isObject, get, has, isArray } from 'lodash-es';
import { type WorkflowVariable } from '@coze-workflow/variable';
import { type DTODefine } from '@coze-workflow/base';

/**
 * Traverse variable references in DTO JSON
 * @param data
 * @param cb
 * @returns
 */
export function traverseRefsInNodeJSON(
  data: unknown,
  cb: (_ref: DTODefine.RefExpression) => void,
) {
  if (isObject(data)) {
    // Determine if it conforms to the structure of ValueExpressionDTO
    if (
      get(data, 'type') === 'ref' &&
      get(data, 'content.source') === 'block-output' &&
      has(data, 'content.blockID') &&
      has(data, 'content.name')
    ) {
      cb(data as unknown as DTODefine.RefExpression);
    }

    Object.entries(data).forEach(([_key, _val]) => {
      traverseRefsInNodeJSON(_val, cb);
    }, {});
    return;
  }

  if (isArray(data)) {
    data.forEach(_item => {
      traverseRefsInNodeJSON(_item, cb);
    });
  }
}

/**
 * variable sort
 * @param variable
 * @returns
 */
export const variableOrder = (name?: string) => {
  const orders = {
    USER_INPUT: 2,
    CONVERSATION_NAME: 1,
  };
  return orders[name ?? ''] || 0;
};

export const sortVariables = (variables: WorkflowVariable[]) => {
  if (!variables) {
    return variables;
  }

  return variables.sort(
    (a, b) =>
      variableOrder(b?.viewMeta?.name) - variableOrder(a?.viewMeta?.name),
  );
};
